crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -156,26 +164,100 @@ where
     }
 
     fn headers_range(&self, range: impl RangeBounds<BlockNumber>) -> ProviderResult<Vec<Header>> {
-        self.database.headers_range(range)
+        let mut headers = Vec::new();
+        let start = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n - 1,
+            Bound::Unbounded => self.canonical_in_memory_state.get_canonical_block_number(),
+        };
+
+        for num in start..=end {
+            if let Some(block_state) = self.canonical_in_memory_state.state_by_number(num) {
+                headers.push(block_state.block().block().header.header().clone());
+            } else {
+                let mut db_headers = self.database.headers_range(num..=end)?;
+                headers.append(&mut db_headers);
+                break;
+            }
+        }
+
+        Ok(headers)
     }
 
     fn sealed_header(&self, number: BlockNumber) -> ProviderResult<Option<SealedHeader>> {
+        if let Some(block_state) = self.canonical_in_memory_state.state_by_number(number) {
+            return Ok(Some(block_state.block().block().header.clone()));
+        }
+
         self.database.sealed_header(number)
     }
 
     fn sealed_headers_range(
         &self,
         range: impl RangeBounds<BlockNumber>,
     ) -> ProviderResult<Vec<SealedHeader>> {
-        self.database.sealed_headers_range(range)
+        let mut sealed_headers = Vec::new();
+        let start = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n - 1,
+            Bound::Unbounded => self.canonical_in_memory_state.get_canonical_block_number(),
+        };
```

Comment:
this pattern repeats a few times, should we extract this into a function instead?

something that accepts a closure for the end unbounded case? or just fetches the canonical block number